samza-core/src/test/java/org/apache/samza/operators/impl/TestWatermarkStates.java

```diff
@@ -0,0 +1,101 @@
+import org.apache.samza.system.WatermarkMessage;
+import org.junit.Test;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+
+import static org.apache.samza.operators.impl.WatermarkStates.WATERMARK_NOT_EXIST;
+
+public class TestWatermarkStates {
+
+  @Test
+  public void testUpdate() {
+    SystemStream input = new SystemStream("system", "input");
+    SystemStream intermediate = new SystemStream("system", "intermediate");
+
+    Set<SystemStreamPartition> ssps = new HashSet<>();
+    SystemStreamPartition inputPartition0 = new SystemStreamPartition(input, new Partition(0));
+    SystemStreamPartition intPartition0 = new SystemStreamPartition(intermediate, new Partition(0));
+    SystemStreamPartition intPartition1 = new SystemStreamPartition(intermediate, new Partition(1));
+    ssps.add(inputPartition0);
+    ssps.add(intPartition0);
+    ssps.add(intPartition1);
+
+    Map<SystemStream, Integer> producerCounts = new HashMap<>();
+    producerCounts.put(intermediate, 2);
+
+    WatermarkStates watermarkStates = new WatermarkStates(ssps, producerCounts);
+    IncomingMessageEnvelope envelope = IncomingMessageEnvelope.buildWatermarkEnvelope(inputPartition0, 5L);
+    boolean updated = watermarkStates.update((WatermarkMessage) envelope.getMessage(),
+        envelope.getSystemStreamPartition());
+    assertTrue(updated);
+    assertEquals(watermarkStates.getWatermark(input), 5L);
+    assertEquals(watermarkStates.getWatermark(intermediate), WATERMARK_NOT_EXIST);
+
+    WatermarkMessage watermarkMessage = new WatermarkMessage(6L, "task 0");
+    updated = watermarkStates.update(watermarkMessage, intPartition0);
+    assertFalse(updated);
+    assertEquals(watermarkStates.getWatermarkPerSSP(intPartition0), WATERMARK_NOT_EXIST);
+    assertEquals(watermarkStates.getWatermark(intermediate), WATERMARK_NOT_EXIST);
+
+    watermarkMessage = new WatermarkMessage(3L, "task 1");
+    updated = watermarkStates.update(watermarkMessage, intPartition0);
+    assertFalse(updated);
+    assertEquals(watermarkStates.getWatermarkPerSSP(intPartition0), 3L);
+    assertEquals(watermarkStates.getWatermark(intermediate), WATERMARK_NOT_EXIST);
+
+    watermarkMessage = new WatermarkMessage(10L, "task 0");
+    updated = watermarkStates.update(watermarkMessage, intPartition1);
+    assertFalse(updated);
+    assertEquals(watermarkStates.getWatermarkPerSSP(intPartition1), WATERMARK_NOT_EXIST);
+    assertEquals(watermarkStates.getWatermark(intermediate), WATERMARK_NOT_EXIST);
+
+    watermarkMessage = new WatermarkMessage(4L, "task 1");
+    updated = watermarkStates.update(watermarkMessage, intPartition1);
+    assertTrue(updated);
+    assertEquals(watermarkStates.getWatermarkPerSSP(intPartition1), 4L);
+    assertEquals(watermarkStates.getWatermark(intermediate), 3L);
+
+    watermarkMessage = new WatermarkMessage(8L, "task 1");
```

Comment:
minor: Would be nice to add some comments here to describe the test case. For example, advance input watermark, advance output watermark, test input watermark remains the same, test output watermark remains the same, etc.

Reply:
Added the comments.